extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java

```diff
@@ -1371,53 +1377,62 @@ private static MultiTenancyStrategy getMultiTenancyStrategy(Optional<String> mul
         if (proxyCache == null) {
             proxyCache = new ProxyCache();
             liveReloadBuildItem.setContextObject(ProxyCache.class, proxyCache);
         }
-        Set<String> changedClasses = Collections.emptySet();
+        Set<String> changedClasses = Set.of();
         if (liveReloadBuildItem.getChangeInformation() != null) {
             changedClasses = liveReloadBuildItem.getChangeInformation().getChangedClasses();
         } else {
             //we don't have class change info, invalidate the cache
             proxyCache.cache.clear();
         }
         //create a map of entity to proxy type
-        PreGeneratedProxies preGeneratedProxies = new PreGeneratedProxies();
         TypePool transformedClassesTypePool = createTransformedClassesTypePool(transformedClassesBuildItem,
                 managedClassAndPackageNames);
+
+        PreGeneratedProxies preGeneratedProxies = new PreGeneratedProxies();
+
         try (ProxyBuildingHelper proxyHelper = new ProxyBuildingHelper(transformedClassesTypePool)) {
+            final ConcurrentLinkedDeque<Future<CachedProxy>> generatedProxyQueue = new ConcurrentLinkedDeque<>();
+            Set<String> proxyInterfaceNames = Set.of(ClassNames.HIBERNATE_PROXY.toString());
+
             for (String managedClassOrPackageName : managedClassAndPackageNames) {
-                CachedProxy result;
                 if (proxyCache.cache.containsKey(managedClassOrPackageName)
                         && !isModified(managedClassOrPackageName, changedClasses, combinedIndex)) {
-                    result = proxyCache.cache.get(managedClassOrPackageName);
+                    CachedProxy proxy = proxyCache.cache.get(managedClassOrPackageName);
+                    generatedProxyQueue.add(CompletableFuture.completedFuture(proxy));
                 } else {
-                    Set<String> proxyInterfaceNames = new TreeSet<>();
-                    proxyInterfaceNames.add(ClassNames.HIBERNATE_PROXY.toString()); //always added
                     if (!proxyHelper.isProxiable(combinedIndex.getClassByName(managedClassOrPackageName))) {
                         // we need to make sure the actual class is proxiable
                         continue;
                     }
-                    final String mappedClass = managedClassOrPackageName;
-                    DynamicType.Unloaded<?> unloaded = proxyHelper.buildUnloadedProxy(mappedClass, proxyInterfaceNames);
-                    result = new CachedProxy(unloaded, proxyInterfaceNames);
-                    proxyCache.cache.put(managedClassOrPackageName, result);
+                    generatedProxyQueue.add(buildExecutor.submit(() -> {
+                        DynamicType.Unloaded<?> unloaded = proxyHelper.buildUnloadedProxy(managedClassOrPackageName,
+                                proxyInterfaceNames);
+                        return new CachedProxy(managedClassOrPackageName, unloaded, proxyInterfaceNames);
```

Comment:
Very nice.

I noticed something odd though - the parameter `proxyInterfaceNames` passed to  `buildUnloadedProxy` is a Set whose content is always the same? IFF that's the case then the `buildUnloadedProxy` could resolve the types in advance and also reuse as a constant.
But it looks like suspicious, like if we lost something during previous changes.